lisa/sut_orchestrator/azure/azure.py

```diff
@@ -444,29 +449,33 @@ def _create_deployment_parameters(
         arm_parameters.location = self._azure_runbook.location
 
         nodes_parameters: List[AzureArmParameterNode] = []
-        for node_runbook in env_runbook.nodes:
-            assert isinstance(node_runbook, schema.NodeSpec)
-            azure_node_runbook = node_runbook.get_extended_runbook(
+        for node_space in requirements:
+            assert isinstance(
+                node_space, schema.NodeSpace
+            ), f"actual: {type(node_space)}"
+            azure_node_runbook = node_space.get_extended_runbook(
                 AzureNodeSchema, field_name=AZURE
             )
-            gallery = AzureArmParameterGallery()
-            node_parameter = AzureArmParameterNode(gallery=gallery)
-            node_parameter.name = f"node-{len(nodes_parameters)}"
-            if azure_node_runbook:
-                if azure_node_runbook.vm_size:
-                    node_parameter.vm_size = azure_node_runbook.vm_size
-                if azure_node_runbook.vhd:
-                    node_parameter.vhd = azure_node_runbook.vhd
-                    node_parameter.gallery = None
-            nodes_parameters.append(node_parameter)
 
             # init node
-            node = environment.nodes.from_spec(node_runbook)
-            node_context = node.get_context(NodeContext)
-            node_context.vm_name = node_parameter.name
-            node_context.username = arm_parameters.admin_username
-            node_context.password = arm_parameters.admin_password
-            node_context.private_key_file = self._runbook.admin_private_key_file
+            nodes = environment.nodes.from_requirement(node_space)
+            for node in nodes:
+                gallery = AzureArmParameterGallery()
+                node_parameter = AzureArmParameterNode(gallery=gallery)
+                node_parameter.name = f"node-{len(nodes_parameters)}"
+                if azure_node_runbook:
+                    if azure_node_runbook.vm_size:
+                        node_parameter.vm_size = azure_node_runbook.vm_size
+                    if azure_node_runbook.vhd:
+                        node_parameter.vhd = azure_node_runbook.vhd
+                        node_parameter.gallery = None
+                nodes_parameters.append(node_parameter)
+
+                node_context = node.get_context(NodeContext)
+                node_context.vm_name = node_parameter.name
```

Comment:
As far as I can tell this is just glue code mapping Azure's Python APIs to our own interface. It doesn't merit close inspection / caution in my opinion.

Reply:
The node parameters are sent to azure deployment template (arm_template.json) for provisioning. The context is kept for whole life of node, so it can be used for deleting node, or special test case requirement like add a data disk.